examples/hubert/dataset/hubert_dataset.py

```diff
@@ -77,57 +88,48 @@ def _get_buckets(
+        bucket_ids = torch.bucketize(torch.tensor(lengths), boundaries)
+        for i in range(bucket_ids.size(0)):
+            bucket_id = int(bucket_ids[i])
             if bucket_id in buckets:
                 buckets[bucket_id].append(i)
             else:
                 buckets[bucket_id] = [i]
         for k in buckets:
-            if self.shuffle:
-                random.shuffle(buckets[k])
             buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
+        buckets = {k: v for k, v in sorted(buckets.items())}
         return buckets
 
-    def __iter__(self) -> Iterator[List[int]]:
-        iter_list = []
+    def _update_iter_list(self) -> None:
+        self.iter_list = []
         total_len = 0
         batch = []
-        if self.max_token_count:
-            for k in self.buckets.keys():
-                for i in range(self.buckets[k].size(0)):
-                    index = self.buckets[k][i]
-                    if total_len > self.max_token_count:
-                        iter_list.append(batch)
-                        batch = [index]
-                        total_len = self.lengths[index]
-                    else:
-                        batch.append(index)
-                        total_len += self.lengths[index]
-        else:
-            for k in self.buckets.keys():
-                for i in range(self.buckets[k].size(0)):
-                    index = self.buckets[k][i]
-                    if total_len == self.batch_size:
-                        iter_list.append(batch)
-                        batch = [index]
-                        total_len = 1
-                    else:
-                        batch.append(index)
-                        total_len += 1
-
-        for batch in iter_list:
-            yield batch
+        max_batch_size = self.max_token_count if self.max_token_count else self.batch_size
+        for k in self.buckets:
+            for i in range(self.buckets[k].size(0)):
+                index = int(self.buckets[k][i])
+                sample_length = self.lengths[index] if self.max_token_count else 1
+                if total_len + sample_length <= max_batch_size:
+                    batch.append(self.indices[index])
+                    total_len += sample_length
+                else:
+                    self.iter_list.append(batch)
+                    batch = [self.indices[index]]
+                    total_len = sample_length
+        if len(batch) > 0 and (self.max_token_count or not self.drop_last):
+            self.iter_list.append(batch)
+
+    def __iter__(self) -> Iterator[List[int]]:
+        if self.shuffle:
+            for k in self.buckets:
+                self.buckets[k] = self.buckets[k][torch.randperm(self.buckets[k].size(0))]
+            self._update_iter_list()
+
+        return iter(self.iter_list)
 
     def __len__(self):
-        return len(self.data_source)
+        return len(self.iter_list)
```

Comment:
i had commented before:

"i'd rather not, since the length isn't guaranteed to be stable and we have no control over the order in which the user calls `__iter__` and `__len__`"

for example, if a user calls `__len__` first and uses the returned length to guide how they access the iterator returned by `__iter__`, they're going to hit problems because the returned length will no longer correspond to the iterator

what would justify including this?